content/en/blog/_posts/2021-09-07-data-duplication-in-data-heavy-k8s-env.md

```diff
@@ -0,0 +1,221 @@
+
+You can also get it from the gcloud command by the Snapshot’s given name:
+
+```terminal
+gcloud compute snapshots --project=staging-eu-castai-vt5hy2 describe postgresql-orders-db-gold | grep id:
+```
+
+## Step 4: Create a development environment for each team
+
+Now I have my Golden Snapshot, which is immutable data. Each team will get a copy 
+of this data, and team members can modify it as they see fit, given that a new EBS/persistent 
+disk will be created for each team.
+
+Below I will add K8s YAML for each namespace, replacing the namespace “sandbox01 -> sandbox42”
+names with yq, sed, or templating tools like Kustomize or using variables in the CI/CD pipeline:
+
+```yaml
+apiVersion: snapshot.storage.k8s.io/v1
+kind: VolumeSnapshotContent
+metadata:
+ name: postgresql-orders-db-sandbox01
+ namespace: sandbox01
+spec:
+ deletionPolicy: Retain
+ driver: storage.csi.cast.ai
+ source:
+   snapshotHandle: 'gcp/projects/staging-eu-castai-vt5hy2/global/snapshots/6645363163809389170'
+   #snapshotHandle: 'snap-07ff83d328c981c98' #for AWS
+ volumeSnapshotRef:
+   kind: VolumeSnapshot
+   name: postgresql-orders-db-snap
+   namespace: sandbox01
+---
+apiVersion: snapshot.storage.k8s.io/v1
+kind: VolumeSnapshot
+metadata:
+ name: postgresql-orders-db-snap
+ namespace: sandbox01
+spec:
+ source:
+   volumeSnapshotContentName: postgresql-orders-db-sandbox01
+   ```
+
+VolumeSnapshotContent (VSC) is not a namespaced K8s object. But I will create an 
+instance of this object for every namespace, so the name needs to be unique (hence the namespace in VSC name).
+
+Now it’s time to replace the driver field with the CSI driver installed in the K8s cluster.
+The CAST AI CSI driver supports all of the major cloud services so this YAML will work on AWS,
+GCP, and probably Azure (though Azure has been lagging on their cloud storage drivers for K8s)
+on the CAST AI K8s cluster.
+
+At this point, I need to use the Retain policy, so that the CSI driver doesn’t try to
+delete my manually created EBS disk snapshot.
+
+For GCP, you will have to build this string by hand - add a full project ID and snapshot ID.
+For AWS, it’s just a plain snapshot ID.
+
+VSC also requires specifying which VolumeSnapshot (VS) will use it, so VSC and VS are
+referencing each other.
+
+Now I can create PersistentVolumeClaim from VS above. It’s important to set this first:
+
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+ name: postgres-pv-claim
+ namespace: sandbox01
+spec:
+ dataSource:
+   kind: VolumeSnapshot
+   name: postgresql-orders-db-snap
+   apiGroup: snapshot.storage.k8s.io
+ accessModes:
+   - ReadWriteOnce
+ resources:
+   requests:
+     storage: 21Gi
+```
+
+Usually, Default StorageClass will have WaitForFirstConsumer. This means that the
+actual Cloud Disk will be created from the Golden Snapshot only when some pod bounds that PVC.
```

Comment:
thanks for calling this out. can you link to more info on `WaitForFirstConsumer`?

Reply:
added link